teps/0015-implicit-mapping.md

```diff
@@ -0,0 +1,658 @@
+  END
+
+  FOR taskParam, taskDefaultValue in implicitTaskParams
+    IF explicitTaskParams[taskParam]
+      continue;
+    ENDIF
+    taskName = results[taskParams]
+    paramValue = specParam[taskParams]
+    IF taskName THEN
+      PipelineTasks[i].Params = append(PipelineTasks[i].Params, Params{taskName.result})
+    ELSEIF paramValue THEN
+      PipelineTasks[i].Params = append(PipelineTasks[i].Params, Params{paramValue})
+    ELSEIF taskDefaultValue == nil
+      return error //Param can not be matched and does not have a task default value
+    ENDIF
+  END
+
+  explicitTaskParams.erase()
+
+END
+```
+
+#### Implicit resource mapping
+
+In accordance with requirement 2, the mutating webhook resolves the `PipelineTask.Resources` field as follows:
+
+```
+INITIALIZATION
+hashmap specResources[resource.Name][resource.Value]
+hashset explicitTaskResources[resource.Name]
+implicitTaskResources[]
+k = PipelineTasks.size()
+m = PipelineTasks[i].Resources.size()
+n = implicitTaskResources.size()
+END INITIALIZATION
+
+FOR EVERY resource IN spec.resources
+  specResources[resource.Name]= resource.Value
+END
+
+FOR i = 0; i < k; i++
+
+  IF PipelineTasks[i].taskSpec == nil THEN
+    Resolve PipelineTasks[i].taskRef
+  ENDIF
+
+  Store implicit task resources in implicitTaskResources[] 
+
+  FOR j = 0; j < n; j++
+    explicitTaskResources.insert(PipelineTask[i].Resources[j])
+  END
+
+  FOR j = 0; j < n; i++;
+    IF explicitTaskResources[PipelineTask[i].Resources[j]]
+      continue
+    ENDIF
+    resourceValue = specResources[implicitTaskResources[j]]
+    IF resourceValue THEN
+      PipelineTasks[i].Resources = append(PipelineTasks[i].Resources, Resource{resourceValue})
+    ELSE
+      return error
+    ENDIF
+  ENDIF
+  END
+
+  explicitResourceParams.erase()
+
+END
+```
+
+### Alternatives
+- If a *taskRef* is used, the empty *taskSpec* field could be populated with the fetched *taskRef*.
+Instead of iterating over an implicit parameter or resource hashmap/array, Tekton can iterate over the parameters and resources in the *taskSpec*.
+Additionally, since the webhook has updated the *taskSpec* field, the reconciler would not need to fetch the *taskRef* on every reconcile loop. 
+This could be considered an attractive option that would require a separate PR.
+Partially because this can happen without implicit mapping as well.
+- The mutating webhook could validate an implicit `Pipeline` specification without altering the `Pipeline` spec.
+This would still require resolving parameters and resources which would additionally need to be performed by the `PipelineRun` reconciler again.
+Considering the amount of duplicate work performed by Tekton this is not considered an attractive option.
+
+
+## Test Plan
+
+- Unit tests for checking that the webhook correctly transforms an implicit `Pipeline` specification into an explicit `Pipeline` specification
+
+## Drawbacks
+
+- Implicit `Pipeline` specifications could be more error prone
```

Comment:
I think drawbacks need more details.
In addition, the potential slowdown in the webhook should be considered as a drawback, imo.